libheif/CMakeLists.txt

```diff
@@ -1,84 +1,106 @@
+    error.cc
+    heif.cc
+    heif_context.cc
+    heif_file.cc
+    heif_image.cc
+    heif_hevc.cc
+    heif_colorconversion.cc
+    heif_plugin_registry.cc
+    heif_plugin.cc
+    ${libheif_headers}
+)
+
+# Needed to find libheif/heif_version.h while compiling the library
+target_include_directories(${PROJECT_NAME} PRIVATE ${PROJECT_BINARY_DIR})
 
-# set(CMAKE_VERBOSE_MAKEFILE on)
+# Propagate include/libheif to consume the headers from other projects
+target_include_directories(${PROJECT_NAME} PUBLIC
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/${PROJECT_NAME}>
+    $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    $<INSTALL_INTERFACE:include>
+)
 
-add_definitions(-DHAVE_VISIBILITY)
-add_definitions(-DLIBHEIF_EXPORTS)
+target_compile_definitions(${PROJECT_NAME}
+    PUBLIC 
+        LIBHEIF_EXPORTS
+        HAVE_VISIBILITY
+)
 
-add_library(${LIBHEIF_LIBRARY_NAME} SHARED ${libheif_sources})
 if(LIBDE265_FOUND)
-  target_link_libraries(${LIBHEIF_LIBRARY_NAME} ${LIBDE265_LIBRARIES} ${LIBDE265_LINKDIR})
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_LIBDE265=1)
+    target_sources(${PROJECT_NAME} PRIVATE
+        heif_decoder_libde265.cc
+        heif_decoder_libde265.h
+    )
+
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${LIBDE265_CFLAGS}")
+
+    if (NOT "${LIBDE265_LIBRARY_DIRS}" STREQUAL "")
+        set(LIBDE265_LINKDIR "-L${LIBDE265_LIBRARY_DIRS}")
+    endif()
+
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${LIBDE265_LIBRARIES} ${LIBDE265_LINKDIR})
+
+    install(FILES heif_decoder_libde265.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
 endif()
+
 if(X265_FOUND)
-  target_link_libraries(${LIBHEIF_LIBRARY_NAME} ${X265_LIBRARIES} ${X265_LINKDIR})
-endif()
-if(CMAKE_COMPILER_IS_GNUCXX)
-  set_target_properties(${LIBHEIF_LIBRARY_NAME} PROPERTIES CXX_VISIBILITY_PRESET hidden)
+    target_compile_definitions(${PROJECT_NAME} PRIVATE HAVE_X265=1)
+    target_sources(${PROJECT_NAME} PRIVATE
+        heif_encoder_x265.cc
+        heif_encoder_x265.h
+    )
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${X265_CFLAGS}")
+
+    if (NOT "${X265_LIBRARY_DIRS}" STREQUAL "")
+        set(X265_LINKDIR "-L${X265_LIBRARY_DIRS}")
+    endif()
+
+    target_link_libraries(${PROJECT_NAME} PRIVATE ${X265_LIBRARIES} ${X265_LINKDIR})
+
+    install(FILES heif_encoder_x265.h DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
 endif()
 
-if(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
-  SET_TARGET_PROPERTIES(${LIBHEIF_LIBRARY_NAME} PROPERTIES COMPILE_FLAGS "-fPIC")
-endif(CMAKE_SYSTEM_PROCESSOR STREQUAL "x86_64")
+write_basic_package_version_file(${PROJECT_NAME}ConfigVersion.cmake COMPATIBILITY ExactVersion)
+
+install(TARGETS ${PROJECT_NAME} EXPORT ${PROJECT_NAME}Config
+    RUNTIME DESTINATION ${CMAKE_INSTALL_BINDIR}
+    LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
+    ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR}
+)
+
+install(FILES ${libheif_headers} DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/${PROJECT_NAME})
```

Comment:
This will install all headers including the various private headers. Public headers are only `heif_cxx.h`, `heif.h`, `heif_plugin.h` and `heif_version.h` (generated). See https://github.com/strukturag/libheif/blob/ae12405f9d9e69791c858f79341355e068302ffe/libheif/Makefile.am#L76

Reply:
Thanks, I will adapt it ASAP